commit-graph: fix buggy --expire-time option

derrickstolee opened the pull request:
This will need a change in microsoft/scalar and microsoft/vfsforgit to handle the correct input. It has _never worked_.

derrickstolee committed:
commit-graph: fix buggy --expire-time option

The commit-graph builtin has an --expire-time option that takes a
datetime using OPT_EXPIRY_DATE(). However, the implementation inside
expire_commit_graphs() was treating a non-zero value as a number of
seconds to subtract from "now".

Update t5323-split-commit-graph.sh to demonstrate the correct value
of the --expire-time option by actually creating a crud .graph file
with mtime earlier than the expire time. Instead of using a super-
early time (1980) we use an explicit, and recent, time. Using
test-tool chmtime to create two files on either end of an exact
second, we create a test that catches this failure no matter the
current time. Using a fixed date is more portable than trying to
format a relative date string into the --expiry-date input.

I noticed this when inspecting some Scalar repos that had an excess
number of commit-graph files. In Scalar, we were using this second
interpretation by using "--expire-time=3600" to mean "delete graphs
older than one hour ago" to avoid deleting a commit-graph that a
foreground process may be trying to load.

Also I noticed that the help text was copied from the --max-commits
option. Fix that help text.

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

derrickstolee commented:
/submit

kewillford commented on t/t5324-split-commit-graph.sh:
```diff
@@ -210,8 +210,10 @@ test_expect_success 'test merge stragety constants' '
 		test_line_count = 2 $graphdir/commit-graph-chain &&
 		test_commit 15 &&
-		git commit-graph write --reachable --split --size-multiple=10 --expire-time=1980-01-01 &&
+		touch -m -t 201801010000.00 $graphdir/extra.graph &&
```
Should we also have a file the is newer than the expire time and make sure it is still there?

derrickstolee replied:
That would be a good way to future-proof it (with an explicit file on each side)

kewillford reviewed (approved):
LGTM one comment on another possible test case.

kewillford commented on commit-graph.c:
```diff
@@ -1707,7 +1707,7 @@ static void expire_commit_graphs(struct write_commit_graph_context *ctx)
 	timestamp_t expire_time = time(NULL);
 
 	if (ctx->split_opts && ctx->split_opts->expire_time)
-		expire_time -= ctx->split_opts->expire_time;
```
One other question here because it seems this should be working since it was taking the current time in seconds and subtracting some number of seconds to create the expire time.  Was the issue just in the test and the fact that it was passing in a date instead of number of seconds?

derrickstolee replied:
The `OPT_EXPIRY_DATE()` parses our seconds as "something" but it wasn't a literal integer value. It likely read our `3600` and thought of it as `3600-01-01 00:00.00`.

derrickstolee committed:
t5319: replace 'touch -m' with 'test-tool chmtime'

The use of 'touch -m' to modify a file's mtime is slightly less
portable than using our own 'test-tool chmtime'. The important
thing is that these pack-files are ordered in a special way to
ensure the multi-pack-index selects some as the "newer" pack-files
when resolving duplicate objects.

Reported-by: Jeff King <peff@peff.net>
Signed-off-by: Derrick Stolee <dstolee@microsoft.com>
